src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs

```diff
@@ -298,11 +319,222 @@ void writeEnumElement(string kind, int value, string operationName, string? extr
+                var baseName = GetSubName(type.Name);
+                WriteLine($"{accessibility} virtual void {GetVisitorName(type)}({type.Name} operation) => DefaultVisit(operation);");
+            }
+
+            Unbrace();
+
+            WriteLine(@"public abstract partial class OperationVisitor<TArgument, TResult>
+    {
+        public virtual TResult Visit(IOperation operation, TArgument argument) => operation is null ? default(TResult) : operation.Accept(this, argument);
+        public virtual TResult DefaultVisit(IOperation operation, TArgument argument) => default(TResult);
+        internal virtual TResult VisitNoneOperation(IOperation operation, TArgument argument) => default(TResult);");
+            Indent();
+
+            foreach (var type in types)
+            {
+                if (type.SkipInVisitor) continue;
+
+                WriteObsoleteIfNecessary(type.Obsolete);
+                var accessibility = type.IsInternal ? "internal" : "public";
+                WriteLine($"{accessibility} virtual TResult {GetVisitorName(type)}({type.Name} operation, TArgument argument) => DefaultVisit(operation, argument);");
+            }
+
+            Unbrace();
+            WriteLine("#endregion");
         }
+
+        private void WriteObsoleteIfNecessary(ObsoleteTag? tag)
+        {
+            if (tag is object)
+            {
+                WriteLine($"[Obsolete({tag.Message}, error: {tag.ErrorText})]");
+            }
+        }
+
+        private string GetVisitorName(Node type)
+        {
+            Debug.Assert(!type.SkipInVisitor, $"{type.Name} is marked as skipped in visitor, cannot generate classes.");
+            return type.VisitorName ?? $"Visit{GetSubName(type.Name)}";
+        }
+
+        private List<Property> GetAllProperties(AbstractNode node)
+        {
+            var properties = node.Properties.ToList();
+
+            AbstractNode? @base = node;
+            while (true)
+            {
+                string baseName = @base.Base;
+                @base = _typeMap[baseName];
+                if (@base is null) break;
+                properties.AddRange(@base.Properties);
+            }
+
+            return properties;
+        }
+
+        private string GetSubName(string operationName) => operationName[1..^9];
+
+        private string GetKind(AbstractNode node)
+        {
+            while (node.OperationKind?.Include == false)
+            {
+                node = _typeMap[node.Base] ??
+                    throw new InvalidOperationException($"{node.Name} is not being included in OperationKind, but has no base type!");
+            }
+
+            if (node.OperationKind?.Entries.Count > 0)
+            {
+                return node.OperationKind.Entries.Where(e => e.EditorBrowsable != false).Single().Name;
+            }
+
+            return GetSubName(node.Name);
+        }
+
+        private bool IsIOperationType(string typeName) => _typeMap.ContainsKey(typeName) ||
```

Comment:
>IsIOperationType [](start = 21, length = 16)

consider making this a local function too

Reply:
I'd prefer to keep these `Is` methods outside, they'll be necessary in other functions later when we generate the OperationCloner.

---
In reply to: [313627735](https://github.com/dotnet/roslyn/pull/37791#discussion_r313627735) [](ancestors = 313627735)